### PR TITLE
GDScript: Fix object iterator opcodes

### LINKED ISSUE #74686: 🐛 Iterators: trying to iterate over `self` terminates the game application with no error message
### Godot version

v4.0.stable.official [92bee43ad]

### System information

Windows 10 Pro

### Issue description

In a class that implements the `iterator` interface, using a `for` loop internally iterating over `self` can cause the game application to crash.

### Context (can be skipped)

I'm working on a deck builder game and last week I started migrating the project to Godot 4.

I use a lot of classes with the gdscript's `iterator` interface to normalize how I iterate over different collections. However, after migrating, all code paths that were iterating over the instance internally using `for _ in self:` started failing silently and closing  or hanging the game application. The error is _not reported_ in the output or debugger panels. All of the classes were working correctly in 3.5.

I decided to create a new project to test the `self` iteration and I have 2 scenes:
- one created from scratch in Godot 4 
- and one with code copied and pasted from my old scripts (which I don't think matters, but I'm adding it for completion sake)

I tried to simplify the copied code as much as possible while still reproducing the error and the project can be found attached.

The weirdest thing to me is that this works: 
```gd
	func get_element_index_v2(element):
		var index = 0

		for e in iterate_slots():
			if element == e:
				return index
			index += 1

		return NOT_FOUND_INDEX


	func iterate_slots():
		return self
```

while this doesn't: 
```gd
	func get_element_index(element):
		var index = 0

		for e in self:
			if element == e:
				return index
			index += 1

		return NOT_FOUND_INDEX
```

------

A similar problem can be seen in the new godot 4 scene:

This works:  
```gd
	func print_it():
		for number in self:
			print(number)
```

while this doesn't:
```gd
	func has_number(num):
		for n in self:
			if n == num:
				return true
		return false
```

### Video demo

https://user-images.githubusercontent.com/89042777/224162415-6a1f8520-3505-4990-9c5d-10271d77c8fb.mp4

### Steps to reproduce

1. Create a class that implements the `iterator` methods
2. Create a new method that iterates over the instance using `self`
3. Add some check inside the for loop, because using only `print` works for some reason
4. Try to run the game, and the app will crash or hang

### Minimal reproduction project

[self-iterator-bugreport.zip](https://github.com/godotengine/godot/files/10935448/self-iterator-bugreport.zip)

### LINKED ISSUE #89290: Infinite loop caused by typed for-loop variable when looping through `PackedDataContainer`
### Tested versions

ffc41fb76df5922321cdd98cce12715a039629b0 - reproducible
4.2.1 - reproducible

### System information

Godot v4.3.dev (55ebf09ea) - Arch Linux #1 SMP PREEMPT_DYNAMIC Sun, 03 Mar 2024 00:30:36 +0000 - Tty - GLES3 (Compatibility) - AMD Radeon RX 6900 XT (radeonsi, navi21, LLVM 17.0.6, DRM 3.57, 6.7.8-arch1-1) () - AMD Ryzen 9 5950X 16-Core Processor (32 Threads)

### Issue description

When looping through a `PackedDataContainer` whose members should only be of type `PackedDataContainerRef`, specifying the latter as the type of a for-loop variable will cause an infinite loop when looping through the elements.

### Steps to reproduce

Run the MRP.

The infinite loop is caused by `ref : PackedDataContainerRef `. Replacing that with just `ref` fixes the issue.

### Minimal reproduction project (MRP)

[TypedPDCRBug.zip](https://github.com/godotengine/godot/files/14541453/TypedPDCRBug.zip)

### LINKED ISSUE #91155: Crash when iterating over custom GDScript iterator captured within lambda
### Tested versions

Reproducible in: 4.3.dev [780e1a504]

### System information

Windows 11 (10.0.22631)

### Issue description

I'm not sure of the correlation here, but it seems that if you in a GDScript script create an instance of a custom iterator, capture that iterator within a lambda to iterate over it, and then invoke that lambda, like so:

```gdscript
extends Node

class RangeIterator:
    var _count: int
    var _current: int

    func _init(count: int) -> void:
        _count = count
        _current = 0

    func should_continue() -> bool:
        return _current < _count

    func _iter_init(arg: Variant) -> bool:
        _current = 0
        return should_continue()

    func _iter_next(arg: Variant) -> bool:
        _current += 1
        return should_continue()

    func _iter_get(arg: Variant) -> Variant:
        return _current

func _init() -> void:
    var iterator := RangeIterator.new(4)

    var iterate := func() -> void:
        for value in iterator:
            pass

    iterate.call()
```

... you will crash the application [here](https://github.com/godotengine/godot/blob/780e1a50408360cf0cf93c0b55b59e9d7b4ad0b1/modules/gdscript/gdscript_vm.cpp#L2997) due to `instruction_args` being `nullptr`, with the following backtrace:

```
GDScriptFunction::call(GDScriptInstance * p_instance, const Variant * * p_args, int p_argcount, Callable::CallError & r_err, GDScriptFunction::CallState * p_state) (modules/gdscript/gdscript_vm.cpp:2997)
GDScriptLambdaCallable::call(const Variant * * p_arguments, int p_argcount, Variant & r_return_value, Callable::CallError & r_call_error) (modules/gdscript/gdscript_lambda_callable.cpp:118)
Callable::callp(const Variant * * p_arguments, int p_argcount, Variant & r_return_value, Callable::CallError & r_call_error) (core/variant/callable.cpp:58)
_VariantCall::func_Callable_call(Variant * v, const Variant * * p_args, int p_argcount, Variant & r_ret, Callable::CallError & r_error) (core/variant/variant_call.cpp:1025)
`_register_variant_builtin_methods'::`2'::Method_Callable_call::call(Variant * base, const Variant * * p_args, int p_argcount, Variant & r_ret, const Vector<Variant> & p_defvals, Callable::CallError & r_error) (core/variant/variant_call.cpp:2064)
Variant::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Variant & r_ret, Callable::CallError & r_error) (core/variant/variant_call.cpp:1226)
GDScriptFunction::call(GDScriptInstance * p_instance, const Variant * * p_args, int p_argcount, Callable::CallError & r_err, GDScriptFunction::CallState * p_state) (modules/gdscript/gdscript_vm.cpp:1746)
GDScript::_create_instance(const Variant * * p_args, int p_argcount, Object * p_owner, bool p_is_ref_counted, Callable::CallError & r_error) (modules/gdscript/gdscript.cpp:182)
GDScript::instance_create(Object * p_this) (modules/gdscript/gdscript.cpp:414)
Object::set_script(const Variant & p_script) (core/object/object.cpp:969)
Node::set_script(const Variant & p_script) (scene/main/node.cpp:3867)
Object::set(const StringName & p_name, const Variant & p_value, bool * r_valid) (core/object/object.cpp:265)
SceneState::instantiate(SceneState::GenEditState p_edit_state) (scene/resources/packed_scene.cpp:343)
PackedScene::instantiate(PackedScene::GenEditState p_edit_state) (scene/resources/packed_scene.cpp:2077)
Main::start() (main/main.cpp:3812)
widechar_main(int argc, wchar_t * * argv) (platform/windows/godot_windows.cpp:179)
_main() (platform/windows/godot_windows.cpp:206)
main(int argc, char * * argv) (platform/windows/godot_windows.cpp:220)
WinMain(HINSTANCE__ * hInstance, HINSTANCE__ * hPrevInstance, char * lpCmdLine, int nCmdShow) (platform/windows/godot_windows.cpp:234)
```

### Steps to reproduce

1. Run the main scene of the MRP
2. Observe the crash

### Minimal reproduction project (MRP)

[captured-iterators.zip](https://github.com/godotengine/godot/files/15110127/captured-iterators.zip)

### PR DESCRIPTION
**Pull request by dalexeev**
* Fixes #74686.
* Fixes #89290.
* Fixes #91155.

See [core methods](https://github.com/godotengine/godot/blob/7abe0c6014022874378cb64a11b26b0f0f178324/core/variant/variant_setget.cpp#L1304-L1924) for reference:

https://github.com/godotengine/godot/blob/7abe0c6014022874378cb64a11b26b0f0f178324/core/variant/variant_setget.cpp#L1304-L1924

### CODE REVIEW

File: modules/gdscript/gdscript_vm.cpp [hunk 1]
```diff
@@ -2649,6 +2649,8 @@ Variant GDScriptFunction::call(GDScriptInstance *p_instance, const Variant **p_a
 				GET_VARIANT_PTR(counter, 0);
 				GET_VARIANT_PTR(container, 1);
 
+				*counter = Variant(); // TODO: Fix in compiler/codegen?
```
Comment (vnen) [inline modules/gdscript/gdscript_vm.cpp:2652]: Please remove the `TODO` comment. I think it's fine doing it here, as doing with the codegen would have to execute yet another instruction.

File: modules/gdscript/gdscript_vm.cpp [hunk 2]
```diff
@@ -2976,6 +2978,9 @@ Variant GDScriptFunction::call(GDScriptInstance *p_instance, const Variant **p_a
 #else
 				Object *obj = *VariantInternal::get_object(container);
 #endif
+
+				*counter = Variant(); // TODO: Fix in compiler/codegen?
```
Comment (vnen) [inline modules/gdscript/gdscript_vm.cpp:2982]: Here too.

**Review by vnen — CHANGES_REQUESTED** [review]
Looks good, just need to remove the `TODO` comments.

**Review by vnen — APPROVED** [review]
LGTM. Thanks!

**Comment by vnen** [comment]
There are some weird errors on the CI tests though, not sure what's going on.

**Comment by dalexeev** [comment]
> There are some weird errors on the CI tests though, not sure what's going on.

* See #91120.

**Comment by dalexeev** [comment]
To fix #74686 and #91155 I added the following:

```diff
@@ -2998,11 +2998,10 @@ Variant GDScriptFunction::call(GDScriptInstance *p_instance, const Variant **p_a
 				VariantInternal::initialize(&vref, Variant::ARRAY);
 				*VariantInternal::get_array(&vref) = ref;
 
-				Variant **args = instruction_args; // Overriding an instruction argument, but we don't need access to that anymore.
-				args[0] = &vref;
+				const Variant *args[] = { &vref };
 
 				Callable::CallError ce;
-				Variant has_next = obj->callp(CoreStringNames::get_singleton()->_iter_init, (const Variant **)args, 1, ce);
+				Variant has_next = obj->callp(CoreStringNames::get_singleton()->_iter_init, args, 1, ce);
 
 #ifdef DEBUG_ENABLED
 				if (ref.size() != 1 || ce.error != Callable::CallError::CALL_OK) {
@@ -3332,11 +3331,10 @@ Variant GDScriptFunction::call(GDScriptInstance *p_instance, const Variant **p_a
 				VariantInternal::initialize(&vref, Variant::ARRAY);
 				*VariantInternal::get_array(&vref) = ref;
 
-				Variant **args = instruction_args; // Overriding an instruction argument, but we don't need access to that anymore.
-				args[0] = &vref;
+				const Variant *args[] = { &vref };
 
 				Callable::CallError ce;
-				Variant has_next = obj->callp(CoreStringNames::get_singleton()->_iter_next, (const Variant **)args, 1, ce);
+				Variant has_next = obj->callp(CoreStringNames::get_singleton()->_iter_next, args, 1, ce);
 
 #ifdef DEBUG_ENABLED
 				if (ref.size() != 1 || ce.error != Callable::CallError::CALL_OK) {
```

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**